### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by aarononeal** [restic/restic#2566]
Output of `restic version`
--------------------------
`restic 0.9.6 compiled with go1.13.4 on linux/amd64`

How did you run restic exactly?
-------------------------------
`restic backup /mnt/drive`

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Not relevant

Expected behavior
-----------------
Backup proceeds without error or warning

Actual behavior
---------------
error: open /mnt/drive/file: input/output error

Steps to reproduce the behavior
-------------------------------
Backup tens of thousands of files from SeaDrive (FUSE). See [Seafile] (https://www.seafile.com/).

Do you have any idea what may have caused this?
-----------------------------------------------
Yes, SeaDrive limits the number of open files and apparently restic is exceeding the limit during backup.

Do you have an idea how to solve the issue?
-------------------------------------------
Options: 
1) Reduce the number of open files 
2) Retry the I/O 
3) Add a parameter to control number of open files or concurrency
4) When I/O fails, note the open file count, reduce the limit by 1, and retry

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Restic is usually pretty awesome!

**Comment by rawtaz** [comment]
Hm. I'm not sure restic can determine that a file operation failed because a third party behind a virtual filesystem is limiting stuff. Is that really something that is amongst restic's concerns to work around? Would arguably be better to fix the root cause instead. But that's just my thought.

**Comment by aarononeal** [comment]
All operating systems have an open file limit. I don't see how this is any different other than it happens to be a low limit. Restic should be resilient to this. Additionally, it seems useful to be able to throttle I/O or concurrency, and that could also address the issue.

**Comment by aarononeal** [comment]
Near as I can tell from looking at the code there is already an internal option for read concurrency and only 2 files are archived concurrently by default. 

https://github.com/restic/restic/blob/d70a4a93509870f382df8f79859eba2319f171f8/internal/archiver/archiver.go#L109

I captured an `strace` log and it is hard to follow in some places but that seems to be what it is constrained to.

I think the problem might be happening during scan and not archive. I don't know why restic opens and then closes files immediately after stat without reading anything. The process eventually hits a point where every file open triggers the below.

```
18778 21:36:21 newfstatat(AT_FDCWD, "/mnt/drive/Shared with groups/library/file.txt",  <unfinished ...>
18778 21:36:21 <... newfstatat resumed> {st_mode=S_IFREG|0644, st_size=8352, ...}, AT_SYMLINK_NOFOLLOW) = 0
18778 21:36:21 openat(AT_FDCWD, "/mnt/drive/Shared with groups/library/file.txt", O_RDONLY|O_NOFOLLOW|O_CLOEXEC <unfinished ...>
18778 21:36:21 <... openat resumed> )   = -1 EIO (Input/output error)
```

Perhaps this is a throttling and not a concurrency issue. For now I'll follow up with the file system vendor, but I am interested to learn why restic is opening files after stat.

**Comment by MichaelEischer** [comment]
`EIO` suggests a problem of the underlying file system. If restic were out of file descriptors the errno should be `[EMFILE]           The process has already reached its limit for open file descriptors.`.

**Comment by aarononeal** [comment]
I don't really understand why restic is opening and then immediately closing every file after stat, but will assume there's a good reason. While that may be what's triggering it, the underlying file system should be able to handle, so I'll close and follow up with vendor.

**Comment by MichaelEischer** [comment]
For the logfile which you posted the two syscalls are triggered from https://github.com/restic/restic/blob/v0.9.6/internal/archiver/archiver.go#L334 and https://github.com/restic/restic/blob/v0.9.6/internal/archiver/archiver.go#L356 .

The `stat` call is required to know whether a file or something else is stored at the path. And the call to `open` fails in the log excerpt. There's actually a missing call to `close` at https://github.com/restic/restic/blob/v0.9.6/internal/archiver/archiver.go#L379 . However, normally the only way to trigger this is if someone swaps the file with something else just inbetween the first call to stat and the open call afterwards.

**Comment by aarononeal** [comment]
Thank you @MichaelEischer! After getting oriented with the code I was able to make some changes to debug and watch various processes during the backup. That's going to be helpful going forward.

You're right, it does look like a `close` is missing, though I was able to confirm it wasn't hitting that path in my case.

My system was showing plenty of available open file slack and I was only seeing this issue when backing up the virtual filesystem. But after experimenting, I determined I was running into the user and not system limit for the user that had the FS mounted. The solution was: https://unix.stackexchange.com/a/145025

    $ vi /etc/security/limits.conf

    * soft     nofile         65535
    * hard     nofile         65535

The open limitation seems to be resolved.

I am a little surprised that Restic's algorithm requires opening every file even when it's determined not to have changed. When dealing with a network FS, particularly one with caching, this can trigger unnecessary I/O.

In any case, thank you all for the assistance! Restic has been great.

**Comment by fd0** [comment]
> The stat call is required to know whether a file or something else is stored at the path. And the call to open fails in the log excerpt. There's actually a missing call to close at https://github.com/restic/restic/blob/v0.9.6/internal/archiver/archiver.go#L379 . However, normally the only way to trigger this is if someone swaps the file with something else just inbetween the first call to stat and the open call afterwards.

@MichaelEischer great catch! Would you mind opening a PR which adds the close (throwing away the error it may return)? :)

**Comment by MichaelEischer** [comment]
@fd0 Please have a look at #2570 . I've just noticed that the test which I've added current fails for the Windows build. I'll take a look at it in the evening.

**Comment by rawtaz** [comment]
I'm reopening this for tracking until #2570 is updated by @MichaelEischer so we can merge it :)
